# Rating Books — o que os números deste acervo dizem

Acervo de **52.478 livros** raspados do Goodreads (25 colunas, 74 MB). A análise responde a quatro
perguntas:

1. Quais gêneros têm mais avaliações no total e melhores notas médias?
2. Quais livros são os mais bem avaliados dentro de cada gênero?
3. Quais autores se destacam (melhor nota média) dentro de cada gênero?
4. Quais os livros são mais caros?

Além dos números, o notebook mostra o caminho até eles: o que o dataset esconde e onde cada
resposta perde confiança.

Os rankings de nota (perguntas 1b, 2 e 3) usam uma **nota ponderada pelo número de avaliações**
(média bayesiana), explicada na seção 2. Sem ela, livros com 2 ou 7 avaliações nota 5,0 ficavam à
frente de clássicos avaliados por centenas de milhares de leitores.

In [1]:

from pathlib import Path

import ast

import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go

IMAGENS = Path("images")
FIGURAS: dict[str, go.Figure] = {}

pd.set_option("display.max_colwidth", 55)


Duas funções ajudam no resto do notebook: `estilo()` deixa todos os gráficos com a mesma
aparência, e `registrar()` guarda cada figura para exportar no final.


In [2]:

AZUL, VERMELHO, VERDE, LARANJA = "#4E79A7", "#E15759", "#59A14F", "#F28E2B"


def estilo(fig, titulo, altura=500):
    fig.update_layout(
        title=dict(text=titulo, font=dict(size=19, color="#1F1F1F"), x=0.01, xanchor="left"),
        font=dict(size=13, color="#444"),
        plot_bgcolor="white",
        paper_bgcolor="white",
        height=altura,
        margin=dict(l=8, r=8, t=62, b=8),
        hoverlabel=dict(font_size=12, bgcolor="white"),
    )
    fig.update_xaxes(gridcolor="#EDEDED", linecolor="#D8D8D8", zeroline=False)
    fig.update_yaxes(gridcolor="#EDEDED", linecolor="#D8D8D8", zeroline=False)
    return fig


def registrar(fig, nome):
    FIGURAS[nome] = fig
    return fig


def miles(valor):
    return f"{valor:,.0f}".replace(",", ".")


## 1. Carregando o acervo

Uso `low_memory=False` para o pandas ler todas as linhas de uma vez. Sem isso, colunas com tipos
misturados podem ser lidas com o tipo errado.


In [3]:

df = pd.read_csv("books_1.Best_Books_Ever.csv", low_memory=False)

print(f"{miles(len(df))} livros x {df.shape[1]} colunas")
print()
df.dtypes.to_string()


52.478 livros x 25 colunas



'bookId                  str\ntitle                   str\nseries                  str\nauthor                  str\nrating              float64\ndescription             str\nlanguage                str\nisbn                    str\ngenres                  str\ncharacters              str\nbookFormat              str\nedition                 str\npages                   str\npublisher               str\npublishDate             str\nfirstPublishDate        str\nawards                  str\nnumRatings            int64\nratingsByStars          str\nlikedPercent        float64\nsetting                 str\ncoverImg                str\nbbeScore              int64\nbbeVotes              int64\nprice                   str'

In [4]:

df[["title", "author", "rating", "numRatings", "bookFormat", "language"]].head(8)


,title,author,rating,numRatings,bookFormat,language
0,The Hunger Games,Suzanne Collins,4.33,6376780,Hardcover,English
1,Harry Potter and the Order of the Phoenix,"J.K. Rowling, Mary GrandPré (Illustrator)",4.50,2507623,Paperback,English
2,To Kill a Mockingbird,Harper Lee,4.28,4501075,Paperback,English
3,Pride and Prejudice,"Jane Austen, Anna Quindlen (Introduction)",4.26,2998241,Paperback,English
4,Twilight,Stephenie Meyer,3.60,4964519,Paperback,English
5,The Book Thief,Markus Zusak (Goodreads Author),4.37,1834276,Hardcover,English
6,Animal Farm,"George Orwell, Russell Baker (Preface), C.M. Woodho...",3.95,2740713,Mass Market Paperback,English
7,The Chronicles of Narnia,"C.S. Lewis, Pauline Baynes (Illustrator)",4.26,517740,Paperback,English


## 2. O que precisa ser arrumado antes de responder

Antes de agrupar, vejo o que está faltando. Duas das quatro perguntas dependem de colunas que não
cobrem o acervo inteiro.


In [5]:

faltando = (df.isna().mean() * 100).round(1).sort_values(ascending=False)
faltando.head(12).to_frame("% ausente").T


,edition,series,firstPublishDate,price,language,publisher,pages,bookFormat,description,publishDate,likedPercent,coverImg
% ausente,90.6,55.3,40.6,27.4,7.3,7.0,4.5,2.8,2.5,1.7,1.2,1.2


A coluna de preço vem como **texto**, não como número. Sem converter, ordenar por preço seria
ordenar alfabeticamente — e `"9.99"` apareceria antes de `"100.00"`.


In [6]:

df["price"] = pd.to_numeric(df["price"], errors="coerce")

com_preco = df["price"].notna()
print(f"dtype apos conversao: {df['price'].dtype}")
print(f"ausentes: {miles(df['price'].isna().sum())} ({100 * df['price'].isna().mean():.1f}%)")
print()
df.loc[com_preco, "price"].describe().round(2).to_frame("price (com valor)").T


dtype apos conversao: float64
ausentes: 14.377 (27.4%)



,count,mean,std,min,25%,50%,75%,max
price (com valor),38101.0,9.66,24.3,0.84,3.24,5.2,8.86,898.64


### Linhas repetidas e livros sem avaliação

Alguns livros aparecem mais de uma vez com o mesmo `bookId` — são a mesma linha raspada duas vezes.
Se ficarem, contam em dobro nas somas e nos rankings. Também há livros com **zero avaliações**: a
nota deles vem como `0`, o que puxaria para baixo qualquer média em que entrassem.

In [7]:
repetidas = df["bookId"].duplicated().sum()
df = df.drop_duplicates(subset="bookId").reset_index(drop=True)

sem_avaliacao = df["numRatings"] == 0

print(f"linhas repetidas removidas: {repetidas}")
print(f"livros restantes          : {miles(len(df))}")
print(f"livros sem avaliacao      : {sem_avaliacao.sum()} (ficam fora das medias de nota)")

linhas repetidas removidas: 54
livros restantes          : 52.424
livros sem avaliacao      : 71 (ficam fora das medias de nota)


### Nota ponderada: por que a nota crua não serve para ranquear

A coluna `rating` é a média das estrelas que cada livro recebeu. O problema é que ela ignora
**quantas pessoas** deram essas estrelas. Metade do acervo tem menos de ~2.300 avaliações, e uma
parte tem tão poucas que a nota é quase sorte: basta o autor e dois amigos darem 5 estrelas.

In [8]:
avaliados = df[~sem_avaliacao]
poucas = avaliados["numRatings"] < 10

print(f"livros com menos de 10 avaliacoes: {miles(poucas.sum())}")
print(f"  ...e com nota 4,5 ou mais       : {miles((poucas & (avaliados['rating'] >= 4.5)).sum())}")
print()
print("Top 10 pela nota crua:")
avaliados.nlargest(10, ["rating", "numRatings"])[["title", "author", "rating", "numRatings"]]

livros com menos de 10 avaliacoes: 2.370
  ...e com nota 4,5 ou mais       : 1.343

Top 10 pela nota crua:


,title,author,rating,numRatings
18852,Bits and Pieces: Tales and Sonnets,Jas T. Ward (Goodreads Author),5.0,36
11414,Traces of My Love: The Secrets of Heart,"C.L. Ross, Ashley McKinley (Editor), Feri Nugroho (...",5.0,35
22870,Unleash the POWER of Branding: The Smart Guide to B...,Joe Praveen Sequeira,5.0,32
20883,Broken Promises: The Suspenseful Sequel To The Nove...,Lashone Vs Shaneek (Goodreads Author),5.0,28
23136,The Right Person,Stacy A. Padula (Goodreads Author),5.0,24
37584,Santa's Little Helper,Ronald Destra (Goodreads Author),5.0,22
17809,"God's Medicine Is Best (herbs, vitamins and minerals)",Linda Wise,5.0,20
10750,Liam: Midsummer's Magic Bonus Book,Emmie Lou Kates (Goodreads Author),5.0,17
11831,Reflecto Girl: Book 1,Violet Plum (Goodreads Author),5.0,17
21100,Brant's Light,F.B. Lee,5.0,17


A solução é a **média bayesiana**, a mesma ideia que o IMDb usa no seu Top 250. Cada livro começa
"emprestando" `m` avaliações com a nota média do acervo inteiro (`C`). Só depois as avaliações reais
dele entram na conta:

$$\text{nota ponderada} = \frac{v \cdot R + m \cdot C}{v + m}$$

- `R` = nota do livro, `v` = número de avaliações dele;
- `C` = nota média de todo o acervo, ponderada pelas avaliações;
- `m` = quanto "peso" a média geral recebe. Usei a **mediana** de avaliações por livro.

Na prática: um livro com poucas avaliações fica perto da média geral, porque ainda não provou nada.
Um livro com centenas de milhares de avaliações fica praticamente com a própria nota. E um livro
com exatamente `m` avaliações fica no meio do caminho entre as duas.

A mesma fórmula vale para gêneros e autores: basta somar as avaliações (`v`) e os "pontos"
(`R × v`) de todos os livros do grupo.

In [9]:
avaliados = df[~sem_avaliacao]

C = (avaliados["rating"] * avaliados["numRatings"]).sum() / avaliados["numRatings"].sum()
M = avaliados["numRatings"].median()


def nota_ponderada(pontos, avaliacoes, m=M, c=C):
    """Média bayesiana. `pontos` = soma de nota x avaliações; `avaliacoes` = soma de avaliações."""
    return (pontos + m * c) / (avaliacoes + m)


df["pontos"] = df["rating"] * df["numRatings"]
df["nota_ponderada"] = nota_ponderada(df["pontos"], df["numRatings"])

print(f"C (nota media do acervo)      : {C:.3f}")
print(f"m (mediana de avaliacoes)     : {miles(M)}")

C (nota media do acervo)      : 4.054
m (mediana de avaliacoes)     : 2.322


O efeito aparece logo no topo do acervo. Pela nota crua, os dez primeiros são livros quase
desconhecidos. Pela nota ponderada, aparecem obras que muita gente leu **e** avaliou bem.

In [10]:
cols = ["title", "author", "rating", "numRatings", "nota_ponderada"]

antes = avaliados.nlargest(10, ["rating", "numRatings"]).index
depois = (
    df.sort_values("nota_ponderada", ascending=False)
    .drop_duplicates(subset=["title", "author"])
    .head(10)
)

comparacao = pd.DataFrame({
    "pela nota crua": df.loc[antes, "title"].str.slice(0, 40).values,
    "aval.": df.loc[antes, "numRatings"].map(miles).values,
    "pela nota ponderada": depois["title"].str.slice(0, 40).values,
    "aval. ": depois["numRatings"].map(miles).values,
}, index=range(1, 11))

comparacao

,pela nota crua,aval.,pela nota ponderada,aval.
1,Bits and Pieces: Tales and Sonnets,36,The Complete Calvin and Hobbes,34.893
2,Traces of My Love: The Secrets of Heart,35,"Harry Potter Boxed Set, Books 1-5 (Harry",52.749
3,Unleash the POWER of Branding: The Smart,32,Words of Radiance,207.369
4,Broken Promises: The Suspenseful Sequel,28,Harry Potter Series Box Set,251.756
5,The Right Person,24,It's a Magical World,25.874
6,Santa's Little Helper,22,Know My Name,55.887
7,"God's Medicine Is Best (herbs, vitamins",20,Harry Potter Collection,30.910
8,Liam: Midsummer's Magic Bonus Book,17,"The Way of Kings, Part 2",13.194
9,Reflecto Girl: Book 1,17,There's Treasure Everywhere,19.726
10,Brant's Light,17,The Authoritative Calvin and Hobbes: A C,20.857


O gráfico mostra o mesmo efeito para o acervo inteiro. Cada livro aparece duas vezes: em cinza, a
nota crua; em azul, a nota ponderada. À esquerda, onde há poucas avaliações, as notas cruas se
espalham de 1 a 5 e a ponderada puxa todas para perto da média. À direita, as duas praticamente
coincidem.

In [11]:
amostra_peso = df[~sem_avaliacao].sample(4000, random_state=7)

fig = go.Figure()
fig.add_scatter(
    x=amostra_peso["numRatings"], y=amostra_peso["rating"], mode="markers",
    name="Nota crua", marker=dict(color="#BBBBBB", size=5, opacity=0.6),
)
fig.add_scatter(
    x=amostra_peso["numRatings"], y=amostra_peso["nota_ponderada"], mode="markers",
    name="Nota ponderada", marker=dict(color=AZUL, size=5, opacity=0.6),
)
fig.add_hline(y=C, line_dash="dot", line_color=VERMELHO,
              annotation_text=f"media do acervo ({C:.2f})", annotation_position="bottom right")
fig.add_vline(x=M, line_dash="dot", line_color="#999",
              annotation_text=f"m = {miles(M)} aval.", annotation_position="top left")

registrar(estilo(fig, "Nota crua x nota ponderada pelo numero de avaliacoes", altura=520),
          "07-nota-crua-x-ponderada")
fig.update_xaxes(type="log", title="Numero de avaliacoes (escala log)")
fig.update_yaxes(title="Nota")
fig.update_layout(legend=dict(orientation="h", y=1.02, x=1, xanchor="right", yanchor="bottom"))
fig.show()

### Gêneros: uma coluna que precisa virar várias linhas

A coluna `genres` guarda uma lista em formato de texto (`['Fantasy', 'Young Adult', ...]`). O
comando `ast.literal_eval` transforma esse texto em lista de verdade, e o `explode` cria uma
linha para cada combinação de livro e gênero.

Isso tem um efeito que aparece mais adiante: cada livro entra em **7,8 gêneros** em média, então
uma mesma avaliação acaba contada em vários gêneros.


In [12]:

df["genres"] = df["genres"].apply(lambda x: ast.literal_eval(x) if pd.notna(x) else [])

df_exploded = df.explode("genres")
todos_generos = pd.Series([g for lista in df["genres"] for g in lista])

print(f"antes do explode : {miles(len(df))} linhas")
print(f"depois do explode: {miles(len(df_exploded))} linhas")
print(f"generos distintos: {todos_generos.nunique()}")
print(f"media de generos por livro: {len(todos_generos) / len(df):.1f}")


antes do explode : 52.424 linhas
depois do explode: 411.961 linhas
generos distintos: 982
media de generos por livro: 7.8


## 3. Pergunta 1 — gêneros com mais avaliações

Para cada gênero, calculo a soma das avaliações, quantos livros distintos existem ali e a nota do
gênero. A nota agora é **ponderada**: soma os pontos (`nota × avaliações`) de todos os livros do
gênero e aplica a média bayesiana. Assim, um livro com 300 mil avaliações pesa mais que um com 2, e
um gênero com pouquíssimas avaliações fica perto da média geral.

In [13]:
genre_stats = df_exploded.groupby("genres").agg(
    total_ratings=("numRatings", "sum"),
    pontos=("pontos", "sum"),
    book_count=("title", "nunique"),
).sort_values("total_ratings", ascending=False)

genre_stats["nota_ponderada"] = nota_ponderada(genre_stats["pontos"], genre_stats["total_ratings"])

tabela = genre_stats.head(15)[["total_ratings", "nota_ponderada", "book_count"]].copy()
tabela["total_ratings"] = tabela["total_ratings"].map(miles)
tabela["nota_ponderada"] = tabela["nota_ponderada"].round(3)

tabela

,total_ratings,nota_ponderada,book_count
genres,,,
Fiction,798.573.850,4.047,30081
Fantasy,397.479.523,4.104,14328
Young Adult,335.370.254,4.092,11362
Audiobook,333.726.811,4.091,7204
Romance,324.048.146,4.026,14527
Adult,291.362.839,4.038,8016
Classics,290.134.692,4.094,6796
Novels,246.286.519,3.985,7724
Contemporary,239.241.847,3.977,10139


In [14]:

top15 = genre_stats.head(15).sort_values("total_ratings").reset_index()

fig = registrar(estilo(px.bar(
    top15, x="total_ratings", y="genres", orientation="h",
    text_auto=".2s", color_discrete_sequence=[AZUL],
), "Generos com mais avaliacoes"), "01-generos-avaliacoes")

fig.update_traces(textposition="outside", textfont_color="#555")
fig.update_layout(xaxis_title="Total de avaliacoes", yaxis_title="")
fig.show()


Fiction, Fantasy e Young Adult lideram em volume. Como cada livro entra, em média, em 7,8
gêneros, esses gêneros acabam se sobrepondo.


### 1b. Melhores notas

Ranqueio os gêneros pela nota ponderada e mantenho o mínimo de 5 livros por gênero. A ponderação
já resolve a maior parte do problema: antes, o topo era **Baha I**, com 6 livros e só 1.588
avaliações somadas. Agora ele cai para perto da média, e sobem gêneros que têm nota alta **e**
volume.

In [15]:
reliable = (
    genre_stats[genre_stats["book_count"] >= 5]
    .sort_values("nota_ponderada", ascending=False)
    .head(15)
)

tabela = reliable[["nota_ponderada", "total_ratings", "book_count"]].copy()
tabela["total_ratings"] = tabela["total_ratings"].map(miles)
tabela["nota_ponderada"] = tabela["nota_ponderada"].round(3)

tabela

,nota_ponderada,total_ratings,book_count
genres,,,
Comic Strips,4.612,716.979,59
Cartoon,4.610,367.823,37
Scripture,4.489,30.108,24
Lds Non Fiction,4.469,161.535,29
Fandom,4.455,288.064,10
Anti Racist,4.422,799.853,46
Manhwa,4.398,14.935,10
Ghana,4.388,190.982,10
Wizards,4.386,68.395,8


In [16]:
top_avg = reliable.sort_values("nota_ponderada").reset_index()

fig = registrar(estilo(px.bar(
    top_avg, x="nota_ponderada", y="genres", orientation="h", text_auto=".2f",
    color="nota_ponderada", color_continuous_scale=[AZUL, VERMELHO], range_color=[4.2, 4.7],
), "Generos com melhores notas (nota ponderada, min. 5 livros)"), "02-generos-nota-media")

fig.update_layout(xaxis_title="Nota ponderada", yaxis_title="", coloraxis_showscale=False)
fig.update_xaxes(range=[4.0, 4.7])
fig.show()

## 4. Pergunta 2 — os mais bem avaliados de cada gênero

Para cada um dos seis gêneros mais populares, pego os três livros de maior **nota ponderada**. Se o
mesmo título aparece em mais de uma edição, fica só a de nota ponderada mais alta.

In [17]:
generos_populares = (
    df_exploded.groupby("genres")["title"].nunique()
    .sort_values(ascending=False).head(6).index.tolist()
)

top3 = (
    df_exploded[df_exploded["genres"].isin(generos_populares)]
    .sort_values("nota_ponderada", ascending=False)
    .drop_duplicates(subset=["genres", "title", "author"])
    .groupby("genres").head(3)
    .copy()
)
top3["genres"] = pd.Categorical(top3["genres"], categories=generos_populares, ordered=True)
top3["posicao"] = top3.groupby("genres", observed=True).cumcount() + 1
top3 = top3.sort_values(["genres", "posicao"])

tabela = top3[["genres", "posicao", "title", "author", "rating", "numRatings", "nota_ponderada"]].copy()
tabela["numRatings"] = tabela["numRatings"].map(miles)
tabela["nota_ponderada"] = tabela["nota_ponderada"].round(3)

tabela

,genres,posicao,title,author,rating,numRatings,nota_ponderada
256,Fiction,1,The Complete Calvin and Hobbes,Bill Watterson,4.82,34.893,4.772
7008,Fiction,2,"Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5)","J.K. Rowling, Mary GrandPré (Illustrator)",4.78,52.749,4.749
635,Fiction,3,Words of Radiance,Brandon Sanderson (Goodreads Author),4.75,207.369,4.742
2699,Romance,1,Mark of the Lion Trilogy,Francine Rivers (Goodreads Author),4.77,11.402,4.649
322,Romance,2,A Court of Mist and Fury,Sarah J. Maas (Goodreads Author),4.62,332.646,4.616
5273,Romance,3,Heartstopper: Volume Two,Alice Oseman (Goodreads Author),4.64,33.942,4.602
7008,Fantasy,1,"Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5)","J.K. Rowling, Mary GrandPré (Illustrator)",4.78,52.749,4.749
635,Fantasy,2,Words of Radiance,Brandon Sanderson (Goodreads Author),4.75,207.369,4.742
409,Fantasy,3,Harry Potter Series Box Set,J.K. Rowling,4.73,251.756,4.724
7008,Young Adult,1,"Harry Potter Boxed Set, Books 1-5 (Harry Potter, #1-5)","J.K. Rowling, Mary GrandPré (Illustrator)",4.78,52.749,4.749


O gráfico organiza esses livros em uma matriz: cada linha é um gênero, cada coluna é a posição,
e a cor é a nota ponderada. Dentro de cada célula estão o título, a nota original e o número de
avaliações. Agora todos os livros do topo têm milhares de avaliações — nenhum chega lá com 7.

In [18]:
notas = top3.pivot(index="genres", columns="posicao", values="nota_ponderada")
celulas = top3.assign(
    texto=(
        top3["title"].str.slice(0, 34)
        + "<br>nota " + top3["rating"].map(lambda v: f"{v:.2f}")
        + "  ·  " + top3["numRatings"].map(miles) + " aval."
    )
).pivot(index="genres", columns="posicao", values="texto")

fig = go.Figure(go.Heatmap(
    z=notas.values,
    x=["1o", "2o", "3o"],
    y=notas.index.astype(str),
    text=celulas.values,
    texttemplate="%{text}",
    textfont=dict(size=11),
    colorscale="Blues",
    zmin=4.3,
    zmax=4.8,
    xgap=4,
    ygap=4,
    colorbar=dict(title="Nota<br>ponderada", len=0.85),
))

fig.update_layout(
    title=dict(text="Livros mais bem avaliados nos generos mais populares (nota ponderada)",
               font=dict(size=19, color="#1F1F1F"), x=0.01, xanchor="left"),
    font=dict(size=13, color="#444"),
    plot_bgcolor="white",
    paper_bgcolor="white",
    height=560,
    margin=dict(l=8, r=8, t=62, b=8),
    xaxis=dict(side="top", title="", showgrid=False),
    yaxis=dict(autorange="reversed", title="", showgrid=False),
)

registrar(fig, "06-livros-por-genero")
fig.show()

## 5. Pergunta 3 — autores que se destacam por gênero

Para cada autor dentro de cada gênero, somo as avaliações e os pontos de todos os livros dele e
aplico a mesma nota ponderada. Continuo exigindo pelo menos 2 livros do autor no gênero.

Antes, a lista era "os 20 maiores recordes em qualquer gênero", e o topo ficava com autores de
poucas avaliações repetidos em vários gêneros (Elias Zapple, Kenneth Thomas). Agora mostro o
**melhor autor em cada um dos 12 gêneros mais populares**, o que deixa a comparação mais justa entre
gêneros.

In [19]:
author_genre = df_exploded.groupby(["genres", "author"]).agg(
    total_ratings=("numRatings", "sum"),
    pontos=("pontos", "sum"),
    book_count=("title", "nunique"),
).reset_index()

author_genre = author_genre[author_genre["book_count"] >= 2].copy()
author_genre["nota_media"] = author_genre["pontos"] / author_genre["total_ratings"]
author_genre["nota_ponderada"] = nota_ponderada(author_genre["pontos"], author_genre["total_ratings"])

generos_autores = (
    df_exploded.groupby("genres")["title"].nunique()
    .sort_values(ascending=False).head(12).index
)
candidatos = author_genre[author_genre["genres"].isin(generos_autores)]

top_authors = (
    candidatos.loc[candidatos.groupby("genres")["nota_ponderada"].idxmax()]
    .set_index("genres").loc[generos_autores].reset_index()
)

tabela = top_authors[["genres", "author", "nota_media", "nota_ponderada", "total_ratings", "book_count"]].copy()
tabela["total_ratings"] = tabela["total_ratings"].map(miles)
tabela[["nota_media", "nota_ponderada"]] = tabela[["nota_media", "nota_ponderada"]].round(3)

tabela

,genres,author,nota_media,nota_ponderada,total_ratings,book_count
0,Fiction,Bill Watterson,4.698,4.694,408.687,15
1,Romance,S.T. Abby,4.649,4.604,28.728,5
2,Fantasy,Masashi Kishimoto,4.589,4.560,40.040,2
3,Young Adult,J.K. Rowling,4.556,4.556,6.257.653,13
4,Contemporary,Shannon Messenger (Goodreads Author),4.651,4.618,39.002,3
5,Nonfiction,Gordon B. Hinckley,4.607,4.566,28.936,3
6,Adult,S.T. Abby,4.649,4.604,28.728,5
7,Novels,Brandon Sanderson (Goodreads Author),4.722,4.681,35.523,3
8,Historical Fiction,Jane Austen,4.571,4.540,35.756,3
9,Mystery,S.T. Abby,4.642,4.573,17.579,3


In [20]:
autores_plot = top_authors.sort_values("nota_ponderada").copy()
autores_plot["autor"] = autores_plot["author"].str.replace(r"\s*\(Goodreads Author\)", "", regex=True)
autores_plot["label"] = autores_plot["genres"] + "  ·  " + autores_plot["autor"].str.slice(0, 30)

fig = registrar(estilo(px.bar(
    autores_plot, x="nota_ponderada", y="label", orientation="h", text_auto=".2f",
    color_discrete_sequence=[VERDE],
), "Autor com melhor nota ponderada nos 12 generos mais populares", altura=620),
    "03-autores-por-genero")

fig.update_layout(xaxis_title="Nota ponderada", yaxis_title="")
fig.update_xaxes(range=[4.0, 4.8])
fig.update_traces(textposition="outside", textfont_color="#555")
fig.show()

## 6. Pergunta 4 — os livros mais caros

Rank direto pelo preço. Vale lembrar que o acervo guarda **edições**, não obras: o mesmo título
aparece como paperback, capa dura ou box set, com preços que não dá para comparar entre si.


In [21]:

expensive = df.dropna(subset=["price"]).nlargest(20, "price")

tabela = expensive[["title", "author", "bookFormat", "rating", "numRatings"]].copy()
tabela.insert(3, "price", expensive["price"].map(lambda v: f"${v:,.2f}"))
tabela["numRatings"] = tabela["numRatings"].map(miles)

tabela


,title,author,bookFormat,price,rating,numRatings
32984,I See by My Outfit,Peter S. Beagle (Goodreads Author),Paperback,$898.64,4.11,379
29560,Margin of Safety: Risk-Averse Value Investing Strat...,Seth A. Klarman,Hardcover,$867.05,4.34,5.404
51278,V/Crying of Lot 49/Gravity's Rainbow,Thomas Pynchon,Hardcover,$811.04,4.41,91
36503,"Men's Garments, 1830-1900: Guide to Pattern Cutting...",R.I. Davis,Paperback,$796.14,4.42,19
11517,Viper's Creed,T.L. Shreffler (Goodreads Author),ebook,$791.43,4.29,2.650
18834,Progeny,Patrick C. Greene (Goodreads Author),Paperback,$777.35,3.72,374
5808,Children of Time,Adrian Tchaikovsky (Goodreads Author),Hardcover,$711.55,4.28,58.124
17663,The Oxford English Dictionary (20 Volume Set),"John Andrew Simpson (Editor), E.S.C. Weiner (Editor)",Hardcover,$653.73,4.72,215
19513,My Year in Oman: An American Experience in Arabia D...,Matthew D. Heines (Goodreads Author),Paperback,$636.27,3.45,11
47866,Revolutionary Letters,Diane di Prima,Paperback,$631.99,4.29,322


In [22]:

caros_plot = expensive.sort_values("price").copy()
caros_plot["label"] = caros_plot["title"].str.slice(0, 48) + "..."

fig = registrar(estilo(px.bar(
    caros_plot, x="price", y="label", orientation="h", text_auto=".0f",
    color_discrete_sequence=[LARANJA],
), "Os 20 livros mais caros do acervo", altura=620), "04-livros-mais-caros")

fig.update_layout(xaxis_title="Preco (US$)", yaxis_title="")
fig.update_traces(textposition="outside", textfont_color="#555")
fig.show()


## 7. Extra — preço, nota e popularidade na mesma figura

Amostra de 3.000 livros com mais de mil avaliações. O eixo x está em escala logarítmica porque as
contagens variam de 1 a 7 milhões, e a cor representa o preço. Dá para ver que a nota se concentra
entre 3,8 e 4,4, e que o preço não muda muito esse padrão.


In [23]:

amostra = df[df["numRatings"] > 1000].sample(3000, random_state=42).copy()

fig = registrar(estilo(px.scatter(
    amostra, x="numRatings", y="rating",
    color=amostra["price"].fillna(0),
    color_continuous_scale="Viridis", opacity=0.6,
    labels={"numRatings": "Numero de avaliacoes", "rating": "Nota media", "color": "Preco (US$)"},
), "Preco x nota x avaliacoes", altura=560), "05-preco-nota-avaliacoes")

fig.update_xaxes(type="log")
fig.update_layout(yaxis_title="Nota media")
fig.show()


In [24]:
com_preco = df["price"].notna()

print(f"livros com preco : {miles(com_preco.sum())} de {miles(len(df))}")
print(f"livros sem preco : {miles((~com_preco).sum())} ({100 * (~com_preco).mean():.1f}%)")
print(f"preco mediano    : ${df['price'].median():.2f}")
print(f"sem nenhuma nota : {sem_avaliacao.sum()} (fora das medias de nota)")

livros com preco : 38.068 de 52.424
livros sem preco : 14.356 (27.4%)
preco mediano    : $5.20
sem nenhuma nota : 71 (fora das medias de nota)


## 8. Gerando as figuras do README

Estas imagens aparecem no README. A exportação usa o **Kaleido**, que aproveita o Chrome já
instalado na máquina.


In [25]:

IMAGENS.mkdir(exist_ok=True)

for nome, figura in FIGURAS.items():
    destino = IMAGENS / f"{nome}.png"
    figura.write_image(destino, width=1100, height=figura.layout.height or 500, scale=2)
    print(f"{destino}  ({destino.stat().st_size / 1024:.0f} KB)")


images/07-nota-crua-x-ponderada.png  (551 KB)


images/01-generos-avaliacoes.png  (150 KB)


images/02-generos-nota-media.png  (150 KB)


images/06-livros-por-genero.png  (307 KB)


images/03-autores-por-genero.png  (179 KB)


images/04-livros-mais-caros.png  (263 KB)


images/05-preco-nota-avaliacoes.png  (639 KB)


## 9. Onde estas respostas merecem desconfiança

Os números estão corretos, mas têm limites que precisam ficar claros:

**A soma por gênero conta a mesma avaliação várias vezes.** Um livro com 8 gêneros soma suas
avaliações 8 vezes, e o total geral fica cerca de 10x maior que o número real de avaliações. Use
os valores da pergunta 1 para comparar gêneros entre si, não como contagem de leitores.

**A nota ponderada depende da escolha de `m`.** Usei a mediana de avaliações por livro. Um `m` maior
favorece ainda mais os livros populares; um menor dá mais espaço a livros de nicho bem avaliados. A
ordem do topo muda pouco, mas não é a única escolha possível.

**Popularidade não é qualidade.** A ponderação evita que 7 avaliações decidam o ranking, mas também
faz com que um livro excelente e pouco conhecido demore a subir. É um critério de **confiança** na
nota, não de mérito literário.

**Box sets e coleções competem com os livros avulsos.** Coleções como *Harry Potter Boxed Set*
recebem notas altas de fãs da série e aparecem no topo de vários gêneros.

**O preço compara edições, não obras.** Box set, paperback e ebook do mesmo título entram na
mesma lista.

**A cobertura é parcial.** Cerca de 27% do acervo não tem preço, e 81% está em inglês, o que
influencia qualquer conclusão sobre gêneros.